asn1/CommonLedgerProtocol.asn

```diff
@@ -0,0 +1,103 @@
+CommonLedgerProtocol
+DEFINITIONS
+AUTOMATIC TAGS ::=
+BEGIN
+
+IMPORTS
+    UInt8,
+    UInt32,
+    UInt64,
+    UInt128,
+    UInt256
+    FROM GenericTypes
+
+    Timestamp
+    FROM InterledgerTypes
+
+    InterledgerProtocolError
+    FROM InterledgerProtocol
+;
+
+-- Zero or more protocolData entries, each with a unique
+-- protocolId, describe protocol-specific information about
+-- unpaid, paid, and conditionally paid requests and results
+-- sent over the CommonLedgerProtocol.
+-- The protocolId of each entry describes the protocol for
+-- interpreting the data. Peers need to establish beforehand
+-- which protocols each understand, and which numbers they
+-- use, which can be agreed ad-hoc,except that protocolId 1
+-- is reserved for the Interledger protocol.
+-- When multiple protocolData items are sent in a request,
+-- only the first one will be decisive in triggering the
+-- response, the rest of the protocolData items are considered
+-- side protocols that piggyback on the first one.
+protocolData ::= SEQUENCE OF SEQUENCE {
+  protocolId UInt8 UNIQUE,
+  data OCTET STRING
+}
+
+-- Ack, Response, and Error are the response types.
+-- When using these in a CLP packet, the requestId should match
+-- the requestId of the request they respond to.
+
+Ack ::= ProtocolData
+Response ::= ProtocolData
+Error ::= SEQUENCE {
```

Comment:
My first instinct was that yes, it makes sense to use only one type of error wrapper, but then I looked at the Interledger error and I don't think it's a good fit here:

* `code` - Fine.
* `name` - Fine.
* `triggeredBy` - Unnecessary - there is only one party that could have triggered the error.
* `forwardedBy` - Unnecessary - these are local errors that can't be forwarded.
* `triggeredAt` - Fine.
* `data` - Maybe unnecessary - perhaps data associated with the error should go in `protocolData` which is more extensible.

So my suggestion would be to just:

``` asn1
Error ::= SEQUENCE {
  code IA5String (SIZE(3)),
  name IA5String,
  triggeredAt Timestamp,
  protocolData ProtocolData
}
```

Where `code` could start with `L` to indicate a **l*ocal or **l*edger error. Whatever codes we use here, they shouldn't overlap with Interledger codes.

> but then again we are expecting them to be relayed.

I disagree. When an error occurs on the ledger layer, the connector should trigger its own error, because nobody else should have to care what ledger they are using. Put differently, the connector's error should look the same no matter what ledger was used.

So if the ledger triggers `L99 - AcmePay Is Not Feeling Like It`, the connector should *not* relay it (as an L99) but rather it should emit its own `T01 - Ledger Unreachable`. They should feel free to include the ledger error in the freeform debug info if they want, because that doesn't force anyone to understand it.